File: src/supervisor/peers/peer-cred.ts
// Provenance: exact pinned source and transformation are recorded in vendor/baby-provenance/file-map.json.
/** Unix socket peer credential lookup via Node-API native SO_PEERCRED. */

import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import type { Socket } from 'node:net';

const require = createRequire(import.meta.url);

interface PeerCredNative {
  getPeerCred(fd: number): { ok: boolean; uid?: number; gid?: number; pid?: number };
}

let native: PeerCredNative | undefined;

function loadNative(): PeerCredNative {
  if (native) return native;
  const candidates = [
    new URL('../../native/peercred/build/Release/peer_cred.node', import.meta.url),
    new URL('../../native/peercred/build/Debug/peer_cred.node', import.meta.url),
    new URL('../../../../src/native/peercred/build/Release/peer_cred.node', import.meta.url),
    new URL('../../../../src/native/peercred/build/Debug/peer_cred.node', import.meta.url),
  ];
  let lastError: unknown;
  for (const candidate of candidates) {
    try {
      native = require(fileURLToPath(candidate)) as PeerCredNative;
      return native;
    } catch (error) {
      lastError = error;
    }
  }
  throw lastError instanceof Error ? lastError : new Error('se-z peer credential addon is unavailable');
}

function getSocketFd(socket: Socket): number | undefined {
  const handle = (socket as unknown as { _handle?: { fd?: number } })._handle;
  if (handle && typeof handle.fd === 'number') {
    return handle.fd;
  }
  return undefined;
}

export function getSocketPeerCred(
  socket: Socket,
): { uid: number; gid: number; pid: number } | undefined {
  const fd = getSocketFd(socket);
  if (fd === undefined) return undefined;
  try {
    const cred = loadNative().getPeerCred(fd);
    if (!cred.ok || cred.uid === undefined || cred.gid === undefined || cred.pid === undefined) {
      return undefined;
    }
    return { uid: cred.uid, gid: cred.gid, pid: cred.pid };
  } catch {
    return undefined;
  }
}

export function getSocketPeerUid(socket: Socket): number | undefined {
  return getSocketPeerCred(socket)?.uid;
}

export function getSocketPeerGid(socket: Socket): number | undefined {
  return getSocketPeerCred(socket)?.gid;
}

File: src/native/peercred/src/peer_cred.cc
// Provenance: exact pinned source and transformation are recorded in vendor/baby-provenance/file-map.json.
#include <napi.h>

#if defined(__linux__)
#include <sys/socket.h>
#include <unistd.h>
#endif

namespace {

struct PeerCredResult {
  int uid = -1;
  int gid = -1;
  int pid = -1;
  bool ok = false;
};

PeerCredResult getPeerCredFromFd(int fd) {
  PeerCredResult result;
#if defined(__linux__) && defined(SO_PEERCRED)
  struct ucred cred {};
  socklen_t len = sizeof(cred);
  if (getsockopt(fd, SOL_SOCKET, SO_PEERCRED, &cred, &len) == 0) {
    result.uid = static_cast<int>(cred.uid);
    result.gid = static_cast<int>(cred.gid);
    result.pid = static_cast<int>(cred.pid);
    result.ok = true;
  }
#endif
  return result;
}

Napi::Value GetPeerCred(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  if (info.Length() < 1 || !info[0].IsNumber()) {
    Napi::TypeError::New(env, "fd (number) required").ThrowAsJavaScriptException();
    return env.Null();
  }

  const int fd = info[0].As<Napi::Number>().Int32Value();
  const PeerCredResult cred = getPeerCredFromFd(fd);
  Napi::Object obj = Napi::Object::New(env);
  obj.Set("ok", Napi::Boolean::New(env, cred.ok));
  if (cred.ok) {
    obj.Set("uid", Napi::Number::New(env, cred.uid));
    obj.Set("gid", Napi::Number::New(env, cred.gid));
    obj.Set("pid", Napi::Number::New(env, cred.pid));
  }
  return obj;
}

}  // namespace

Napi::Object Init(Napi::Env env, Napi::Object exports) {
  exports.Set("getPeerCred", Napi::Function::New(env, GetPeerCred));
  return exports;
}

NODE_API_MODULE(peer_cred, Init)
